### PR TITLE
WIP: Change subnet.AvailabilityZone in state from a string to []string and rename.

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

Change subnet.AvailabilityZone in state from a string to []string and rename.  
Update tests accordingly.  Added an upgrade state.  Migration piece in the 
works.

## QA steps

TBD

### CODE REVIEW

File: state/linklayerdevices_ipaddresses.go [hunk 1]
```diff
@@ -222,7 +222,7 @@ func (addr *Address) Remove() (err error) {
 	removeOp := removeIPAddressDocOp(addr.doc.DocID)
 	ops := []txn.Op{removeOp}
 	if addr.ProviderID() != "" {
-		op := addr.st.networkEntityGlobalKeyRemoveOp("address", corenetwork.Id(addr.ProviderID()))
+		op := addr.st.networkEntityGlobalKeyRemoveOp("address", addr.ProviderID())
```
Comment (hmlanigan) [inline state/linklayerdevices_ipaddresses.go]: golinter said was unnecessary

File: provider/gce/environ_network.go [hunk 2]
```diff
@@ -125,9 +125,8 @@ func (e *environ) getInstanceSubnets(
 	for _, iface := range ifaces {
 		if subnetIds.Include(string(iface.ProviderSubnetId)) {
 			results = append(results, makeSubnetInfo(
-				// TODO (manadart 2019-07-11): Move InterfaceInfo to core and remove casting.
-				corenetwork.Id(iface.ProviderSubnetId),
-				corenetwork.Id(iface.ProviderNetworkId),
+				iface.ProviderSubnetId,
```
Comment (hmlanigan) [inline provider/gce/environ_network.go]: golinter said was unnecessary

Comment (manadart) [reply provider/gce/environ_network.go]: I obviously missed some of these that I added before deciding to move the `Id` type into `core/network`.
Thanks for the catch.

File: state/linklayerdevices_ipaddresses.go (hunk 1 above)
Comment (manadart) [reply state/linklayerdevices_ipaddresses.go]: Same again; thanks for fixing.

File: state/spaces.go [hunk 3]
```diff
@@ -62,14 +62,14 @@ func (s *Space) ProviderId() network.Id {
 }
 
 // Subnets returns all the subnets associated with the Space.
-func (s *Space) Subnets() (results []*Subnet, err error) {
-	defer errors.DeferredAnnotatef(&err, "cannot fetch subnets")
```
Comment (hmlanigan) [inline state/spaces.go:57]: drive by clean up

File: apiserver/common/networkingcommon/shims.go [hunk 4]
```diff
@@ -18,67 +18,33 @@ import (
 
 // subnetShim forwards and adapts state.Subnets methods to BackingSubnet.
 type subnetShim struct {
-	subnet *state.Subnet
+	*state.Subnet
```
Comment (hmlanigan) [inline apiserver/common/networkingcommon/shims.go:21]: drive by clean up

File: apiserver/common/networkingcommon/subnets.go [hunk 5]
```diff
@@ -418,24 +420,8 @@ func ListSubnets(api NetworkBacking, args params.SubnetsFilters) (results params
 			)
 			continue
 		}
-		// TODO(babbageclunk): make the empty string a valid space
```
Comment (hmlanigan) [inline apiserver/common/networkingcommon/subnets.go:419]: drive by clean up, removing duplicated code.

File: apiserver/common/networkingcommon/types.go [hunk 6]
```diff
@@ -132,21 +130,27 @@ func BackingSubnetToParamsSubnet(subnet BackingSubnet) params.Subnet {
...
-		VLANTag:    vlantag,
-		ProviderId: string(providerid),
-		Zones:      zones,
-		Status:     status,
-		SpaceTag:   spaceTag.String(),
-		Life:       subnet.Life(),
+		CIDR:              cidr,
+		VLANTag:           vlantag,
+		ProviderId:        string(providerid),
+		ProviderNetworkId: string(providerNetworkid),
```
Comment (hmlanigan) [inline apiserver/common/networkingcommon/types.go:149]: Added ProviderNetworkId when removing duplicated code from ListSubnets()

File: apiserver/common/networkingcommon/types.go [hunk 7]
```diff
@@ -132,21 +130,27 @@ func BackingSubnetToParamsSubnet(subnet BackingSubnet) params.Subnet {
 	cidr := subnet.CIDR()
 	vlantag := subnet.VLANTag()
 	providerid := subnet.ProviderId()
+	providerNetworkid := subnet.ProviderNetworkId()
 	zones := subnet.AvailabilityZones()
 	status := subnet.Status()
-	var spaceTag names.SpaceTag
+
+	// TODO(babbageclunk): make the empty string a valid space
```
Comment (hmlanigan) [inline apiserver/common/networkingcommon/types.go:137]: Relocated when removing duplicated code from ListSubnets()

File: core/network/subnet.go [hunk 8]
```diff
@@ -59,15 +59,6 @@ type SubnetInfo struct {
 	FanInfo *FanCIDRs
 }
 
-// AvailabilityZone returns the first of any availability zones,
```
Comment (hmlanigan) [inline core/network/subnet.go]: No longer needed with change to Subnet.

File: state/spaces_test.go [hunk 9]
```diff
@@ -44,7 +44,7 @@ func (s *SpacesSuite) makeSubnetInfosForCIDRs(c *gc.C, CIDRs []string) []network
 		infos[i] = network.SubnetInfo{
 			CIDR:              cidr,
 			VLANTag:           79,
-			AvailabilityZones: []string{"AvailabilityZone"},
+			AvailabilityZones: []string{"AvailabilityZones"},
```
Comment (hmlanigan) [inline state/spaces_test.go:47]: superfluous changes, Zone to Zones got away from me.

File: state/migration_export.go [hunk 10]
```diff
@@ -1177,15 +1177,10 @@ func (e *exporter) subnets() error {
...
+			AvailabilityZones: subnet.AvailabilityZones(),
 			FanLocalUnderlay:  subnet.FanLocalUnderlay(),
 			FanOverlay:        subnet.FanOverlay(),
 		}
-		// TODO(babbageclunk): at the moment state.Subnet only stores
-		// one AZ.
-		az := subnet.AvailabilityZone()
-		if az != "" {
-			args.AvailabilityZones = []string{az}
-		}
```
Comment (jameinel) [inline state/migration_export.go:1178]: So for migrations, have they always supported zones as a list and we just only ever transferred a single one? (so no compatibility issues with changing the stream?)

File: state/upgrades_test.go [hunk 11]
```diff
@@ -3390,6 +3390,50 @@ func (s *upgradesSuite) TestAddSpaceIdToSpaceDocs(c *gc.C) {
 	s.assertUpgradedData(c, AddSpaceIdToSpaceDocs, expectUpgradedData{col, expected})
 }
 
+func (s *upgradesSuite) TestChangeSubnetAZtoSlice(c *gc.C) {
```
Comment (jameinel) [inline state/upgrades_test.go:3393]: have we tested this manually?

**Review by jameinel — APPROVED** [review]
Overall seems very good, a few hopefully small comments.

**Pull request closed without merging**